base_layer/core/src/base_node/sync/block_sync/synchronizer.rs

```diff
@@ -0,0 +1,243 @@
+        if tip_header.height <= local_metadata.height_of_longest_chain() {
+            debug!(
+                target: LOG_TARGET,
+                "Blocks already synchronized to height {}.", tip_header.height
+            );
+            return Ok(());
+        }
+
+        let tip_hash = tip_header.hash();
+        let best_height = local_metadata.height_of_longest_chain();
+        let best_block = self
+            .blockchain_state
+            .get_header(best_height)
+            .await?
+            .ok_or_else(|| BlockSyncError::ExpectedHeaderNotFound(best_height))?;
+        let best_full_block_hash = best_block.hash();
+        debug!(
+            target: LOG_TARGET,
+            "Starting block sync from peer `{}`. Current best block is #{} `{}`. Syncing to #{} ({}).",
+            peer,
+            best_height,
+            best_full_block_hash.to_hex(),
+            tip_header.height,
+            tip_hash.to_hex()
+        );
+        let request = SyncBlocksRequest {
+            start_hash: best_full_block_hash.clone(),
+            // To the tip!
+            end_hash: tip_hash,
+        };
+
+        let mut block_stream = client.sync_blocks(request).await?;
+        let mut prev_hash = best_full_block_hash;
+        let mut current_header = None;
+        while let Some(block) = block_stream.next().await {
+            let block = block?;
+
+            let header = self
+                .blockchain_state
+                .get_header_by_hash(block.hash.clone())
+                .await?
+                .ok_or_else(|| {
+                    BlockSyncError::ReceivedInvalidBlockBody("Peer sent hash for block header we do not have".into())
+                })?;
+
+            let header_hash = header.hash();
+
+            debug!(
+                target: LOG_TARGET,
+                "Validating and adding block body for block: #{}", header.height
+            );
+
+            if header.prev_hash != prev_hash {
+                return Err(BlockSyncError::PeerSentBlockThatDidNotFormAChain {
+                    expected: prev_hash.to_hex(),
+                    got: header.prev_hash.to_hex(),
+                });
+            }
+
+            let block_body = block
+                .body
+                .map(AggregateBody::try_from)
+                .ok_or_else(|| BlockSyncError::ReceivedInvalidBlockBody("Block body was empty".to_string()))?
+                .map_err(BlockSyncError::ReceivedInvalidBlockBody)?;
+
+            prev_hash = header.hash();
+
+            debug!(
+                target: LOG_TARGET,
+                "Storing block body #{} (PoW = {}, {} input(s), {} output(s), {} kernel(s))",
+                header.height,
+                header.pow_algo(),
+                block_body.inputs().len(),
+                block_body.outputs().len(),
+                block_body.kernels().len(),
+            );
+
+            let (header, block_body) = self.validate_block(header, block_body).await?;
```

Comment:
If the blocks matches the header we had, but the  block fails validation, e.g. due to the chain balance being wrong, how  does this  recover? I'm worried that the Header sync part of  this process is not able to determine that chains including this header  hash cannot  sync and it will  repeatedly try to sync  to them. 

Perhaps  a bad header hash list can  prevent  that header from being synced? Or otherwise maybe we should  put all of the fields that could cause  validation to fail due to consensus in the header (e.g. total  sums  etc) and the only way this part of the process can fail is when a body cannot be  provided  that matches the value committed in the headers. If no valid body (or full block) can be found for a header from any peers, then that header should be marked as poison

Reply:
I think the solution may be to keep track of peer's that provided the incorrect information and avoid them (unless they are the only forced sync peer, in which case there is not much we can do except try that node again).

If a block body sync fails, the state machine will return to the header sync which can select another peer, rewind the state and resync headers. As long as we don't sync from that same peer. If all our peers are providing bad data, then I think all we can do is keep banning and trying someone else.